diff --git a/pcbnew/router/pns_line.cpp b/pcbnew/router/pns_line.cpp
index 8eb1f9352d..8df22729f4 100644
--- a/pcbnew/router/pns_line.cpp
+++ b/pcbnew/router/pns_line.cpp
@@ -33,6 +33,9 @@
 
 #include <geometry/shape_rect.h>
 
+extern bool have_pads;
+extern bool have_vias;
+
 namespace PNS {
 
 LINE::LINE( const LINE& aOther )
@@ -555,7 +558,6 @@ bool LINE::Walkaround( const SHAPE_LINE_CHAIN& aObstacle, SHAPE_LINE_CHAIN& aPat
 const SHAPE_LINE_CHAIN SEGMENT::Hull( int aClearance, int aWalkaroundThickness, int aLayer ) const
 {
     /*DEBUG_DECORATOR* debugDecorator = ROUTER::GetInstance()->GetInterface()->GetDebugDecorator();
-    
     PNS_DBG( debugDecorator, Message, wxString::Format( wxT( "seghull %d %d" ), aWalkaroundThickness, aClearance ) );
     PNS_DBG(debugDecorator, AddShape, &m_seg, RED, 0, wxT("theseg") );
         */
@@ -1021,6 +1023,101 @@ void LINE::dragSegment45( const VECTOR2I& aP, int aIndex )
         }
     }
 
+    /* Henry Dang: If there is no any pads or vias in the selected track
+       then we can remove the stupid segment as well */
+    if(have_pads == false && have_vias == false) {
+        if( aIndex == 0 ) {
+            int len = m_line.Segment(0).Length();
+
+            VECTOR2I A =  m_line.Segment(0).A;
+            VECTOR2I B =  m_line.Segment(0).B;
+            VECTOR2I ip;
+
+            // Check for dragging along x axis
+            if ((A.x == B.x) && (A.y != B.y)) {
+                if((m_line.PointCount() >=3) && (best.Segment(0).B.x == m_line.Segment(1).B.x)){
+                    best.Remove(0, 0);
+                }else if(A.y < B.y){
+                    if(best.PointCount() > 3) {
+                        ip = VECTOR2I(best.Segment(1).B.x, best.Segment(1).B.y - len);
+                        best.Replace(0, 0, ip);
+                    }else{
+                        ip = VECTOR2I(best.Segment(0).B.x, best.Segment(0).B.y - len);
+                        best.Replace(0, 0, ip);
+                    }
+                }else if(A.y > B.y){
+                    if(best.PointCount() > 3) {
+                        ip = VECTOR2I(best.Segment(1).B.x, best.Segment(1).B.y + len);
+                        best.Replace(0, 0, ip);
+                    }else{
+                        ip = VECTOR2I(best.Segment(0).B.x, best.Segment(0).B.y + len);
+                        best.Replace(0, 0, ip);
+                    }
+
+                }
+            }
+            // Check for dragging along y axis
+            if((A.y == B.y) && (A.x != B.x)) {
+                if((m_line.PointCount() >=3) && (best.Segment(0).B.y == m_line.Segment(1).B.y)){
+                    best.Remove(0, 0);
+                }else if(A.x < B.x){
+                    if(best.PointCount() > 3) {
+                        ip = VECTOR2I(best.Segment(1).B.x - len, best.Segment(1).B.y);
+                        best.Replace(0, 0, ip);
+                    }else{
+                        ip = VECTOR2I(best.Segment(0).B.x - len, best.Segment(0).B.y);
+                        best.Replace(0, 0, ip);
+                    }
+                }else if(A.x > B.x){
+
+                    if(best.PointCount() > 3) {
+                        ip = VECTOR2I(best.Segment(1).B.x + len, best.Segment(1).B.y);
+                        best.Replace(0, 0, ip);
+                    }else{
+                        ip = VECTOR2I(best.Segment(0).B.x + len, best.Segment(0).B.y);
+                        best.Replace(0, 0, ip);
+                    }
+                }
+            }
+
+        } else if( aIndex == m_line.SegmentCount() - 1 ) {
+            int len = m_line.Segment(-1).Length();
+
+            VECTOR2I A =  m_line.Segment(-1).A;
+            VECTOR2I B =  m_line.Segment(-1).B;
+            VECTOR2I ip;
+
+            // Check for dragging along x axis
+            if ((A.x == B.x) && (A.y != B.y)) {
+                if((m_line.PointCount() >=3) && (best.Segment(0).B.x == m_line.Segment(-2).A.x)){
+                    best.Remove(-1, -1);
+                }else if(A.y < B.y){
+                    ip = VECTOR2I(best.Segment(0).B.x, best.Segment(0).B.y + len);
+                    best.Remove(-1, -1);
+                    best.Append(ip);
+                }else if(A.y > B.y){
+                    ip = VECTOR2I(best.Segment(0).B.x, best.Segment(0).B.y - len);
+                    best.Remove(-1, -1);
+                    best.Append(ip);
+                }
+            }
+            // Check for dragging along y axis
+            if((A.y == B.y) && (A.x != B.x)) {
+                if((m_line.PointCount() >=3) && (best.Segment(0).B.y == m_line.Segment(-2).A.y)){
+                    best.Remove(-1, -1);
+                }else if(A.x < B.x){
+                    ip = VECTOR2I(best.Segment(0).B.x + len, best.Segment(0).B.y);
+                    best.Remove(-1, -1);
+                    best.Append(ip);
+                }else if(A.x > B.x){
+                    ip = VECTOR2I(best.Segment(0).B.x - len, best.Segment(0).B.y);
+                    best.Remove(-1, -1);
+                    best.Append(ip);
+                }
+            }
+        }
+    }
+
     if( m_line.PointCount() == 1 )
         m_line = best;
     else if( aIndex == 0 )
diff --git a/pcbnew/router/router_tool.cpp b/pcbnew/router/router_tool.cpp
index 2cec48dc74..a7c5350d9b 100644
--- a/pcbnew/router/router_tool.cpp
+++ b/pcbnew/router/router_tool.cpp
@@ -81,6 +81,9 @@ using namespace std::placeholders;
 
 #include <pcb_io/kicad_sexpr/pcb_io_kicad_sexpr.h>
 
+bool have_pads;
+bool have_vias;
+
 using namespace KIGFX;
 
 /**
@@ -2140,6 +2143,43 @@ int ROUTER_TOOL::InlineDrag( const TOOL_EVENT& aEvent )
         return 0;
     }
 
+    /* Henry Dang: Find the selected track and check if there is any pads or vias on it */
+    BOARD_CONNECTED_ITEM* con_item = static_cast<BOARD_CONNECTED_ITEM*>( selection.Front() );
+    TRACKS tracks = board()->TracksInNet(con_item->GetNetCode());
+    PAD* pad;
+
+    have_pads = false;
+
+    for (const PCB_TRACK* track : tracks) {
+        if (track) {
+            if(track->IsSelected()){
+                pad = board()->GetPad(track, ENDPOINT_START);
+                if(pad){
+                    have_pads = true;
+                }
+                pad = board()->GetPad(track, ENDPOINT_END);
+                if(pad){
+                    have_pads = true;
+                }
+            }
+        }
+    }
+
+    std::shared_ptr<CONNECTIVITY_DATA>  pConnectedData = board()->GetConnectivity();
+    std::vector<PCB_VIA*> vias;
+    std::vector<PAD*> pads;
+    pConnectedData->GetConnectedPadsAndVias(con_item, &pads, &vias );
+
+    have_vias = false;
+
+    for (const PCB_VIA* via : vias) {
+        if(via){
+            if(via->GetNetCode() == con_item->GetNetCode()){
+                have_vias = true;
+            }
+        }
+    }
+
     std::set<FOOTPRINT*> footprints;
 
     if( item->Type() == PCB_FOOTPRINT_T )
